### PR TITLE
Make batch matrix multiplication on GPU tunable

### PR DESCRIPTION
**Pull request by t-vi**
Here is a PR making the hardcoded splits in the GPU batch matrix
multiplication tunable through autotvm.

(This is primarily aimed at the AMD GPU backend and done as part
of a project for AMD, but should work for all users of the GPU
schedule like the CUDA/nvptx backend as well.)

I get reasonably good results for the workloads I tried - on par to 1.5x
slower for batch matmul, but quite a speedup when integrated into
BERT.

This is my first PR adding autotvm tunings to the public repo, so please bear with me.

### CODE REVIEW

**Comment by t-vi** [comment]
@tqchen @comaniac This is related the [(batch) matmul discussion on the forums](https://discuss.tvm.ai/t/optimizing-matrix-multiplication-for-gpu/4212/18) you participated in.
@masahi

File: python/tvm/relay/op/strategy/cuda.py [hunk 1]
```diff
@@ -463,9 +463,9 @@ def batch_matmul_strategy_cuda(attrs, inputs, out_type, target):
     """batch_matmul cuda strategy"""
     strategy = _op.OpStrategy()
     strategy.add_implementation(
-        wrap_compute_batch_matmul(topi.nn.batch_matmul),
+        wrap_compute_batch_matmul(topi.cuda.batch_matmul),
         wrap_topi_schedule(topi.cuda.schedule_batch_matmul),
-        name="batch_matmul.cuda",
+        name="batch_matmul.gpu",
```
Comment (comaniac) [inline python/tvm/relay/op/strategy/cuda.py:468]: Why change to `gpu`?

File: topi/python/topi/cuda/batch_matmul.py [hunk 2]
```diff
@@ -14,13 +14,24 @@
...
+import tvm
+from tvm import autotvm
 from tvm import te
 from tvm.contrib import cublas
+from tvm.autotvm.task.space import SplitEntity, OtherOptionEntity
+from .. import nn
 from ..util import traverse_inline, get_const_tuple, get_max_power2_factor
 
-def schedule_batch_matmul(outs):
+@autotvm.register_topi_compute("batch_matmul.gpu")
```
Comment (comaniac) [inline topi/python/topi/cuda/batch_matmul.py:27]: We use `XXX.cuda` in all other places so I'd suggest changing to `batch_matmul.cuda`.

File: topi/python/topi/cuda/batch_matmul.py [hunk 3]
```diff
@@ -14,13 +14,24 @@
...
+def batch_matmul(cfg, x, y):
+    """Compute conv2d with NCHW layout"""
+    return nn.batch_matmul(x, y)
+
+
+@autotvm.register_topi_schedule("batch_matmul.gpu")
```
Comment (comaniac) [inline topi/python/topi/cuda/batch_matmul.py:33]: ditto.

File: topi/python/topi/cuda/batch_matmul.py [hunk 4]
```diff
@@ -51,55 +62,73 @@ def _schedule(op):
...
+        cfg.define_split("tile_x", x, num_outputs=3)
+        cfg.define_knob("auto_unroll_max_step", [8, 16, 32, 64])
+        target = tvm.target.Target.current()
+        if target.target_name in ['nvptx', 'rocm']:
+            # llvm-based backends cannot do non-explicit unrolling
+            cfg.define_knob("unroll_explicit", [1])
+        else:
+            cfg.define_knob("unroll_explicit", [0, 1])
+
+        if cfg.is_fallback:
```
Comment (comaniac) [inline topi/python/topi/cuda/batch_matmul.py:76]: Missing `unroll_explicit` in fallback schedule. Based on your comment, this might result in a problem at LLVM-based backends.

File: topi/python/topi/cuda/batch_matmul.py [hunk 5]
```diff
@@ -51,55 +62,73 @@ def _schedule(op):
...
-        ty, yi = s[C].split(y, nparts=y_nthreads)
-        tx, xi = s[C].split(x, nparts=x_nthreads)
-        thread_x = te.thread_axis((0, x_nthreads), "threadIdx.x")
-        thread_y = te.thread_axis((0, y_nthreads), "threadIdx.y")
+
+        cfg.define_split("tile_y", y, num_outputs=3)
+        cfg.define_split("tile_x", x, num_outputs=3)
+        cfg.define_knob("auto_unroll_max_step", [8, 16, 32, 64])
+        target = tvm.target.Target.current()
+        if target.target_name in ['nvptx', 'rocm']:
```
Comment (comaniac) [inline topi/python/topi/cuda/batch_matmul.py:70]: It reminds me that you may also need to add batch_matmul op strategy to ROCM; otherwise ROCM will never use this schedule.

File: python/tvm/relay/op/strategy/cuda.py (hunk 1 above)
Comment (t-vi) [reply python/tvm/relay/op/strategy/cuda.py:468]: I don't insist on it, but we recently (with the great(!) new softmax schedule) had some confusion about cuda schedules being used on non-cuda.

Comment (comaniac) [reply python/tvm/relay/op/strategy/cuda.py:468]: Yeah I realized that ROCM uses many CUDA schedules and it results in confusion. We can find lots of CUDA schedules have `if target == 'rocm'`. However, it would be better to have a separate PR to correct this if everone agrees to do so.

File: topi/python/topi/cuda/batch_matmul.py (hunk 5 above)
Comment (t-vi) [reply topi/python/topi/cuda/batch_matmul.py:70]: See, one might easily think so. But I know that ROCm uses the schedule. My impression is that calling the shared schedules GPU would make that more clear.

File: topi/python/topi/cuda/batch_matmul.py (hunk 4 above)
Comment (t-vi) [reply topi/python/topi/cuda/batch_matmul.py:76]: Good catch, thank you!

File: topi/python/topi/cuda/batch_matmul.py (hunk 5 above)
Comment (comaniac) [reply topi/python/topi/cuda/batch_matmul.py:70]: Yeah I totally agree. That's why I suggested having another RFC/PR to fix all cases. On the other hand, simply specify `gpu` in `topi/cuda` and cover both CUDA and ROCM are also confusion to me.

File: topi/python/topi/cuda/batch_matmul.py (hunk 4 above)
Comment (t-vi) [reply topi/python/topi/cuda/batch_matmul.py:76]: Upon looking at this, I should use unroll explicit, not just define it.
What I'm unsure about is whether I need to define it in the fallback - conv2d_direct doesn't define it and it seems to work well.

Comment (comaniac) [reply topi/python/topi/cuda/batch_matmul.py:76]: Hmm it's interesting. It means this parameter is never effective in those schedules...others like conv2d_direct uses it like:

`s[output].pragma(kernel_scope, 'unroll_explicit', cfg['unroll_explicit'].val)`

I think you can also fix them if you prefer to.

**Comment by t-vi** [comment]
> Could you provide some evaluation results for this tunable template? For example, the tuning space size, the best performance after 3,000 trials (or less if the tuning space is smaller than 3,000), and the performance of fallback config.

After 2000 trials I get to speed roughly on par with PyTorch (using rocblas) on a Radeon VII for both configurations mentioned in the thread linked about. On a GTX 1080TI, I'm getting to 1.5x-ish results (so 1.5x slower than PyTorch),  down from "a lot more", even if I didn't measure as much.
For the end-to-end bert that I'm running with a few upcoming PRs, I'm going from not competitive performance to significantly faster than vanilla PyTorch on the first example from the PyTorch JIT tutorial.

File: topi/python/topi/cuda/batch_matmul.py (hunk 5 above)
Comment (t-vi) [reply topi/python/topi/cuda/batch_matmul.py:70]: Yeah, I'm sticking to cuda for now to not create two confusions where there was only one. :slightly_smiling_face:

File: python/tvm/relay/op/strategy/cuda.py (hunk 1 above)
Comment (t-vi) [reply python/tvm/relay/op/strategy/cuda.py]: Yeah, I changed it to CUDA for now.

File: topi/python/topi/cuda/batch_matmul.py (hunk 4 above)
Comment (t-vi) [reply topi/python/topi/cuda/batch_matmul.py:76]: I think it should be OK now.

Comment (icemelon9) [reply topi/python/topi/cuda/batch_matmul.py:76]: I think you still need to add `"unroll_explicit"` in the fallback cfg.

Comment (comaniac) [reply topi/python/topi/cuda/batch_matmul.py:76]: Sorry just found that the `unroll_explicit` is still missing in fallback. Also would you mind moving L103 (define `tile_k`) up together with other tuning parameters?

Comment (t-vi) [reply topi/python/topi/cuda/batch_matmul.py:76]: As much fun as geeking out over this is
- I don't think I need the fallback because
  - This schedule works without,
  - other schedules work without,
  - define_knob sets the fallback to the first option by setting `cfg._entity_map` which is what is queried by `__getitem__`.

I cannot move the definition of `tile_k` to before k is defined.

Comment (comaniac) [reply topi/python/topi/cuda/batch_matmul.py:76]: If the first two points matter, we should just remove that parameter to make the tuning space more efficient. Meanwhile, I appreciate the third point that I didn't realize before.

For `tile_k`, I think you should be able to move the definition of `k` up as well. It should be safe because the tuning parameter must be static so it won't depend on other parameters. In this way, we can also put the fallback configs together to make it clearer.

Comment (t-vi) [reply topi/python/topi/cuda/batch_matmul.py:78]: I moved the k and `tile_k` it appears to work and looks much prettier indeed. Thank you for insisting on it.

Comment (t-vi) [reply topi/python/topi/cuda/batch_matmul.py:78]: What else do you think you need to happen?

**Comment by tqchen** [comment]
Thanks @t-vi  for the contribution, and @comaniac for reviewing!
